HeterogeneousCore/AlpakaInterface/interface/prefixScan.h

```diff
@@ -0,0 +1,235 @@
+#ifndef HeterogeneousCore_AlpakaInterface_interface_prefixScan_h
+#define HeterogeneousCore_AlpakaInterface_interface_prefixScan_h
+// Active version
+
+#include <alpaka/alpaka.hpp>
+
+#include "HeterogeneousCore/AlpakaInterface/interface/config.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/workdivision.h"
+#include "FWCore/Utilities/interface/CMSUnrollLoop.h"
+namespace cms {
+  namespace alpakatools {
+
+    template <typename T, typename = std::enable_if_t<std::is_integral_v<T>>>
+    constexpr bool isPowerOf2(T v) {
+      // returns true iif v has only one bit set.
+      while (v) {
+        if (v & 1)
+          return !(v >> 1);
+        else
+          v >>= 1;
+      }
+      return false;
+    }
+
+    template <typename TAcc, typename T, typename = std::enable_if_t<alpaka::isAccelerator<TAcc>>>
+    ALPAKA_FN_ACC ALPAKA_FN_INLINE void warpPrefixScan(const TAcc& acc, int32_t laneId, T const* ci, T* co, uint32_t i) {
+      // ci and co may be the same
+      auto x = ci[i];
+      CMS_UNROLL_LOOP
+      for (int32_t offset = 1; offset < alpaka::warp::getSize(acc); offset <<= 1) {
+        // Force the exact type for integer types otherwise the compiler will find the template resolution ambiguous.
+        using dataType = std::conditional_t<std::is_floating_point_v<T>, T, std::int32_t>;
+        T y = alpaka::warp::shfl(acc, static_cast<dataType>(x), laneId - offset);
+        if (laneId >= offset)
+          x += y;
+      }
+      co[i] = x;
+    }
+
+    template <typename TAcc, typename T, typename = std::enable_if_t<alpaka::isAccelerator<TAcc>>>
+    ALPAKA_FN_ACC ALPAKA_FN_INLINE void warpPrefixScan(const TAcc& acc, int32_t laneId, T* c, uint32_t i) {
+      warpPrefixScan(acc, laneId, c, c, i);
+    }
+
+    // limited to warpSize² elements
+    template <typename TAcc, typename T>
+    ALPAKA_FN_ACC ALPAKA_FN_INLINE void blockPrefixScan(
+        const TAcc& acc, T const* ci, T* co, int32_t size, T* ws = nullptr) {
+      if constexpr (!requires_single_thread_per_block_v<TAcc>) {
+        const auto warpSize = alpaka::warp::getSize(acc);
+        int32_t const blockDimension(alpaka::getWorkDiv<alpaka::Block, alpaka::Threads>(acc)[0u]);
+        int32_t const blockThreadIdx(alpaka::getIdx<alpaka::Block, alpaka::Threads>(acc)[0u]);
```

Comment:
```suggestion
        const int32_t blockDimension = alpaka::getWorkDiv<alpaka::Block, alpaka::Threads>(acc)[0u];
        const int32_t blockThreadIdx = alpaka::getIdx<alpaka::Block, alpaka::Threads>(acc)[0u];
```
for consistency with the previous line.